text/0000-wire-reform.md

```diff
@@ -0,0 +1,146 @@
+
+# No-goals
+
+* This reform does not change the wire decorator syntax.
+* This reform does not change the wire adapter API for lightning platform (we should be able to keep that intact).
+
+# Proposal
+
+This reform is focused on the refactor of the wire decorator code, and the wire-service code. As part of the separation process, there are certain responsibilities that must be well defined:
+
+## Responsibilities of the wire decorator
+
+* To install a prototype descriptor to handle the wired field value from the `vm` of the component.
+* To create an instance of an adapter and link it to the host.
+* To signal to the adapter instance when the component is connected or disconnected.
+* To signal to the adapter instance when the config have changed by providing the new config object.
+* To extract the config value from the host object by relying on the compiler's wire metadata.
+* To signal to the adapter instance when a context is available by providing the new context value. 
+
+## Responsibilities of the wire service
+
+* To define the wire adapter protocol.
+* To provide a referral implementation of the wire adapter protocol.
+* To provide an optional abstraction for existing "legacy" wire adapter factories.
+
+## Implementation Details
+
+* `@lwc/engine` does not know about `@lwc/wire-service` and vice versa.
+* `@lwc/engine` will install a descriptor on the prototype for every wired field during the decorators registration routine.
+* `@wire` decorator will create an instance of the provided `WireAdapter` during the component initialization routine by providing the data callback, and the context callback as arguments. 
+* `@wire` decorator will invoke the adapter's `connect` and `disconnect` based on the internal hooks used per component.
+* `@wire` decorator will create a mutation tracking phase to track any access executed during the computation of the config before calling the `update` routine to be able to detect mutations on those values and issue another update on the adapter instance.
+* `@wire` decorator will implement the logic to provide contextual information when requested by the wire adapter.
+* `@lwc/compiler` will provide a config function per `@wire()` declaration to produce a new config object when invoked with the `component` as the first argument. The `@wire` adapter can rely on that config function to produce a new config object at will.
+* `@lwc/wire-service` becomes lightning platform specific for most part (`register()` method), since anyone can implement the wire adapter protocol.
+
+### Wire Adapter Protocol
+
+The formalization of the wire adapter protocol is important because that enables the interoperability aspect of this feature. The adapter's code should not be aware for the component system, or even the application framework, it only cares about very specific hints to produce a stream of data. The following describes the proposed protocol:
```

Comment:
```suggestion
The formalization of the wire adapter protocol is important because that enables the interoperability aspect of this feature. The adapter's code should not be aware of the component system, or even the application framework. It only cares about very specific hints to produce a stream of data. The following describes the proposed protocol:
```